alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE DeriveGeneric #-}
+
+module Cardano.Ledger.Alonzo.Scripts
+  ( Tag (..),
+    Script (..),
+    ExUnits (..),
+  )
+where
+
+import Cardano.Binary
+import Cardano.Ledger.ShelleyMA.Timelocks
+import Data.Coders
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks)
+
+-- | Marker indicating the part of a transaction for which this script is acting
+-- as a validator.
+data Tag
+  = -- | Validates spending a script-locked UTxO
+    Input
+  | -- | Validates minting new tokens
+    Mint
+  | -- | Validates certificate transactions
+    Cert
+  | -- | Validates withdrawl from a reward account
+    Wdrl
+  deriving (Eq, Generic, Ord, Show)
+
+instance NoThunks Tag
+
+data Script era
+  = NativeScript (Timelock era)
+  | NonNativeScript
+
+-- | Arbitrary execution unit in which we measure the cost of scripts.
+data ExUnits = ExUnits
+  { exUnitsMem :: !Word64,
+    exUnitsSteps :: !Word64
+  }
+  deriving (Eq, Show, Ord)
+
```

Comment:
we need to have a discussion about how we actually handle this execution units tracking. What is in the spec now (ie this structure) is kind of middle ground between two approaches not finalized and needs to change. @WhatisRT we should schedule a meeting for some time in the next week or two